packages/query/src/graphql/plugins/QueryDepthLimitPlugin.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import type {ApolloServerPlugin} from 'apollo-server-plugin-base';
+import {
+  GraphQLSchema,
+  Kind,
+  GraphQLError,
+  DocumentNode,
+  DefinitionNode,
+  printError,
+  ValidationContext,
+  TypeInfo,
+  SelectionNode,
+  FragmentDefinitionNode,
+  OperationDefinitionNode,
+  ASTNode,
+} from 'graphql';
+
+export function validateQueryDepth(maxDepth: number, context: ValidationContext): GraphQLError[] {
+  const errors: GraphQLError[] = [];
+
+  const {definitions} = context.getDocument();
+  const fragments = getFragments(definitions);
+  const operations = getQueriesAndMutations(definitions);
+
+  for (const operation of operations) {
+    if (operation.name && operation.name.value === 'IntrospectionQuery') {
+      continue; // Skip the rest of the loop for this iteration
+    }
+    const depth = determineDepth(operation, fragments, 0, maxDepth, context);
+    if (depth > maxDepth) {
+      errors.push(new GraphQLError(`Query is too deep: ${depth}. Maximum depth allowed is ${maxDepth}.`, [operation]));
+    }
+  }
+  return errors;
+}
+
+function isOperationDefinitionNode(node: DefinitionNode): node is OperationDefinitionNode {
+  return node.kind === Kind.OPERATION_DEFINITION;
+}
+function isFragmentDefinitionNode(node: DefinitionNode): node is FragmentDefinitionNode {
+  return node.kind === Kind.FRAGMENT_DEFINITION;
+}
+
+function getFragments(definitions: readonly DefinitionNode[]): Record<string, FragmentDefinitionNode> {
+  return definitions.filter(isFragmentDefinitionNode).reduce((frags, def) => {
+    frags[def.name.value] = def;
+    return frags;
+  }, {});
+}
+
+function getQueriesAndMutations(definitions: readonly DefinitionNode[]): OperationDefinitionNode[] {
+  return definitions.filter(isOperationDefinitionNode);
+}
+
+function determineDepth(
```

Comment:
I think this should exit when it reaches the limit, there's no point in continuing to recurse if the limit is reached